editoast/src/views/work_schedules.rs

```diff
@@ -0,0 +1,262 @@
+use std::result::Result as StdResult;
+use thiserror::Error;
+use utoipa::ToSchema;
+
+use crate::error::InternalError;
+use crate::error::Result;
+use crate::modelsv2::work_schedules::WorkSchedule;
+use crate::modelsv2::work_schedules::WorkScheduleGroup;
+use crate::modelsv2::work_schedules::WorkScheduleType;
+use crate::modelsv2::Changeset;
+use crate::modelsv2::Create;
+use crate::modelsv2::Model;
+use crate::schema::TrackRange;
+use crate::DbPool;
+
+crate::routes! {
+    "/work_schedules" => {
+        create,
+    }
+}
+
+editoast_common::schemas! {
+    WorkSchedule,
+    WorkScheduleCreateForm,
+    WorkScheduleGroup,
+    WorkScheduleItemForm,
+    WorkScheduleType,
+}
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "work_schedule")]
+pub enum WorkScheduleError {
+    #[error("Name '{name}' already used")]
+    #[editoast_error(status = 400)]
+    NameAlreadyUsed { name: String },
+}
+
+pub fn map_diesel_error(e: InternalError, name: impl AsRef<str>) -> InternalError {
+    if e.message.contains(
+        r#"duplicate key value violates unique constraint "work_schedule_group_name_key""#,
+    ) {
+        WorkScheduleError::NameAlreadyUsed {
+            name: name.as_ref().to_string(),
+        }
+        .into()
+    } else {
+        e
+    }
+}
+
+#[derive(Serialize, Derivative, ToSchema)]
+#[derivative(Default)]
+struct WorkScheduleItemForm {
+    pub start_date_time: NaiveDateTime,
+    pub end_date_time: NaiveDateTime,
+    pub track_ranges: Vec<TrackRange>,
+    pub obj_id: String,
+    pub work_schedule_type: WorkScheduleType,
+}
+
+impl<'de> Deserialize<'de> for WorkScheduleItemForm {
+    fn deserialize<D>(deserializer: D) -> StdResult<WorkScheduleItemForm, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(deny_unknown_fields)]
+        struct Internal {
+            start_date_time: NaiveDateTime,
+            end_date_time: NaiveDateTime,
+            track_ranges: Vec<TrackRange>,
+            obj_id: String,
+            work_schedule_type: WorkScheduleType,
+        }
+        let internal = Internal::deserialize(deserializer)?;
+
+        // Check dates
+        if internal.start_date_time > internal.end_date_time {
+            return Err(SerdeError::custom(format!(
+                "The work_schedule start date '{}' must be before the end date '{}'",
+                internal.start_date_time, internal.end_date_time
+            )));
+        }
+
+        Ok(WorkScheduleItemForm {
+            start_date_time: internal.start_date_time,
+            end_date_time: internal.end_date_time,
+            track_ranges: internal.track_ranges,
+            obj_id: internal.obj_id,
+            work_schedule_type: internal.work_schedule_type,
+        })
+    }
+}
+
+impl WorkScheduleItemForm {
+    pub fn into_work_schedule_changeset(
+        self,
+        work_schedule_group_id: i64,
+    ) -> Result<Changeset<WorkSchedule>> {
```

Comment:
It feels like this function always returns `Ok` and never `Err`. Maybe you could change the signature to remove `Result` ?

Reply:
You're right, thanks :pray: